src/libraries/Native/Unix/System.Native/pal_signal.c

```diff
@@ -17,33 +17,179 @@
+static volatile TerminalInvalidationCallback g_terminalInvalidationCallback = NULL;
+// Callback invoked for SIGCHLD
+static volatile SigChldCallback g_sigChldCallback = NULL;
+static volatile bool g_sigChldConsoleConfigurationDelayed;
+static void (*g_sigChldConsoleConfigurationCallback)(void);
+// Callback invoked for for SIGTTOU while terminal settings are changed.
+static volatile ConsoleSigTtouHandler g_consoleTtouHandler;
+
+// Callback invoked for PosixSignal handling.
+static PosixSignalHandler g_posixSignalHandler = NULL;
+// Tracks whether there are PosixSignal handlers registered.
+static volatile bool g_hasPosixSignalRegistrations[NSIG];
+
 static int g_signalPipe[2] = {-1, -1}; // Pipe used between signal handler and worker
 
-static struct sigaction* OrigActionFor(int sig)
+static bool TryConvertSignalCodeToPosixSignal(int signalCode, PosixSignal* posixSignal)
+{
+    assert(posixSignal != NULL);
+
+    switch (signalCode)
+    {
+        case SIGHUP:
+            *posixSignal = PosixSignalSIGHUP;
+            return true;
+
+        case SIGINT:
+            *posixSignal = PosixSignalSIGINT;
+            return true;
+
+        case SIGQUIT:
+            *posixSignal = PosixSignalSIGQUIT;
+            return true;
+
+        case SIGTERM:
+            *posixSignal = PosixSignalSIGTERM;
+            return true;
+
+        case SIGCHLD:
+            *posixSignal = PosixSignalSIGCHLD;
+            return true;
+
+        case SIGWINCH:
+            *posixSignal = PosixSignalSIGWINCH;
+            return true;
+
+        case SIGCONT:
+            *posixSignal = PosixSignalSIGCONT;
+            return true;
+
+        case SIGTTIN:
+            *posixSignal = PosixSignalSIGTTIN;
+            return true;
+
+        case SIGTTOU:
+            *posixSignal = PosixSignalSIGTTOU;
+            return true;
+
+        case SIGTSTP:
+            *posixSignal = PosixSignalSIGTSTP;
+            return true;
+
+        default:
+            *posixSignal = signalCode;
+            return false;
+    }
+}
+
+int32_t SystemNative_GetPlatformSignalNumber(PosixSignal signal)
 {
-    switch (sig)
+    switch (signal)
+    {
+        case PosixSignalSIGHUP:
+            return SIGHUP;
+
+        case PosixSignalSIGINT:
+            return SIGINT;
+
+        case PosixSignalSIGQUIT:
+            return SIGQUIT;
+
+        case PosixSignalSIGTERM:
+            return SIGTERM;
+
+        case PosixSignalSIGCHLD:
+            return SIGCHLD;
+
+        case PosixSignalSIGWINCH:
+            return SIGWINCH;
+
+        case PosixSignalSIGCONT:
+            return SIGCONT;
+
+        case PosixSignalSIGTTIN:
+            return SIGTTIN;
+
+        case PosixSignalSIGTTOU:
+            return SIGTTOU;
+
+        case PosixSignalSIGTSTP:
+            return SIGTSTP;
+    }
+
+    if (   signal > 0
+        && signal <= NSIG      // Ensure we stay within the static arrays.
+#ifdef SIGRTMAX
+        && signal <= SIGRTMAX  // Runtime check for highest value.
```

Comment:
Might need to check `SIGRTMIN` as well. IIRC, some libcs don't expose the full range of real time signals that the kernel supports because they use a few of them internally.